WalletWasabi.Tests/UnitTests/Crypto/ZeroKnowledge/KnowledgeOfRepTests.cs

```diff
@@ -29,9 +30,9 @@ public void End2EndVerificationSimple(uint scalarSeed1, uint scalarSeed2)
 			var publicPoint = secrets * generators;
 
 			var statement = new Statement(publicPoint, generators);
-			var random = new MockRandom();
-			random.GetBytesResults.Add(new byte[32]);
-			var proof = ProofSystemHelpers.Prove(statement, secrets, random);
+			var mockRandom = new Mock<WasabiRandom>();
+			mockRandom.Setup(rnd => rnd.GetBytes(32)).Returns(new byte[32]);
```

Comment:
This line is very fragile actually. When you change GetBytes(32) -> GetBytes(31), the test still passes.

```csharp
			mockRandom.Setup(rnd => rnd.GetBytes(32)).Returns(new byte[32]).Verifiable();
			var proof = ProofSystemHelpers.Prove(statement, secrets, mockRandom.Object);
			Assert.True(ProofSystemHelpers.Verify(statement, proof));

			mockRandom.Verify();
```

This seems more robust as `mockRandom.Verify();` actually verifies that your mockRandom.Setup was called.

Reply:
Then your test is a black box and your PR does not make any sense.

Reply:
We have to test black boxes, yes.  This is because we have to verify **what** that the SUT does.  Behavioral testing verify **how** the SUT does it.  These two alternatives have nothing to do with the library, we could do behavioral testing with our PM mocks without any problem. Nevertheless you will not find BT in Wasabi.

Anyway, I switched the mock to strict mode.

Reply:
> We have to test black boxes, yes.

No. Your test is a black box. Not the tested code. This is simply because when you let anybody read this test, he will rightly assume that `mockRandom.Setup(rnd => rnd.GetBytes(32)).Returns(new byte[32])` actually does something or is needed for some reason.

Is it needed? No. You can delete the line and the test runs just fine. That's very wrong because it is confusing.

So either we will use Moq strict mode everywhere or we can just revert to the old behavior because it was actually better.